File: gateway/api/v1/approvals.py
"""Approval management endpoints — all require an admin bearer session."""

from fastapi import APIRouter, Depends, HTTPException, Query

from gateway.api.deps import get_approval_manager, require_admin
from gateway.approval.manager import ApprovalManager
from gateway.models.approval import ApprovalResult, ApprovalStatus, ApprovalSummary
from gateway.models.identity import CallerIdentity

router = APIRouter()


@router.get("/", response_model=list[ApprovalSummary])
async def list_approvals(
    status: str | None = Query(
        default=None,
        description="Filter by status: PENDING, APPROVED, DENIED, EXPIRED, USED",
    ),
    limit: int = Query(default=50, le=200),
    manager: ApprovalManager = Depends(get_approval_manager),
    admin: CallerIdentity = Depends(require_admin),
) -> list[ApprovalSummary]:
    """List approval requests, optionally filtered by status."""
    status_filter = ApprovalStatus(status) if status else None
    return await manager.list_requests(
        status_filter=status_filter,
        limit=limit,
        org_id=admin.org_id,
    )


@router.get("/{token}", response_model=ApprovalResult)
async def get_approval(
    token: str,
    manager: ApprovalManager = Depends(get_approval_manager),
    admin: CallerIdentity = Depends(require_admin),
) -> ApprovalResult:
    """Fetch the current status of an approval token."""
    try:
        return await manager.check_token(token, expected_org_id=admin.org_id)
    except ValueError as exc:
        raise HTTPException(status_code=404, detail=str(exc)) from exc


@router.post("/{token}/approve", response_model=ApprovalResult)
async def approve(
    token: str,
    note: str = Query(default=""),
    manager: ApprovalManager = Depends(get_approval_manager),
    admin: CallerIdentity = Depends(require_admin),
) -> ApprovalResult:
    """Approve a pending approval token."""
    try:
        return await manager.approve(
            token,
            admin.caller_id,
            note,
            expected_org_id=admin.org_id,
        )
    except ValueError as exc:
        raise HTTPException(status_code=404, detail=str(exc)) from exc


@router.post("/{token}/deny", response_model=ApprovalResult)
async def deny(
    token: str,
    note: str = Query(default=""),
    manager: ApprovalManager = Depends(get_approval_manager),
    admin: CallerIdentity = Depends(require_admin),
) -> ApprovalResult:
    """Deny a pending approval token."""
    try:
        return await manager.deny(
            token,
            admin.caller_id,
            note,
            expected_org_id=admin.org_id,
        )
    except ValueError as exc:
        raise HTTPException(status_code=404, detail=str(exc)) from exc

File: ui/src/api/approvals.ts
import { apiClient } from './client'

export interface ApprovalSummary {
  token: string
  caller_id: string
  org_id: string
  tool_name: string
  server: string
  status: 'PENDING' | 'APPROVED' | 'DENIED' | 'EXPIRED' | 'USED'
  scope: 'EXECUTION' | 'OUTPUT_RELEASE'
  created_at: string
  expires_at: string
  approver_id: string | null
  decided_at: string | null
}

export async function fetchPendingApprovals(): Promise<ApprovalSummary[]> {
  const response = await apiClient.get<ApprovalSummary[]>('/v1/approvals/', {
    params: { status: 'PENDING' },
  })
  return response.data
}

export async function approveRequest(token: string, note = 'Approved via dashboard'): Promise<void> {
  await apiClient.post(`/v1/approvals/${token}/approve`, null, {
    params: { note },
  })
}

export async function denyRequest(token: string, note = 'Denied via dashboard'): Promise<void> {
  await apiClient.post(`/v1/approvals/${token}/deny`, null, {
    params: { note },
  })
}
